lightning/src/util/events.rs

```diff
@@ -380,9 +381,32 @@ pub trait MessageSendEventsProvider {
 	fn get_and_clear_pending_msg_events(&self) -> Vec<MessageSendEvent>;
 }
 
-/// A trait indicating an object may generate events
+/// A trait indicating an object may generate events.
+///
+/// Events are processed by a handler given to [`process_pending_events`]. Therefore, implementors
+/// should be mindful to avoid lock reentrancy if a handler may call back into the provider.
+///
+/// [`process_pending_events`]: Self::process_pending_events
 pub trait EventsProvider {
-	/// Gets the list of pending events which were generated by previous actions, clearing the list
-	/// in the process.
-	fn get_and_clear_pending_events(&self) -> Vec<Event>;
+	/// Processes any events generated since the last call using the given event handler.
+	///
+	/// Subsequent calls must only process new events. However, handlers must be capable of handling
+	/// duplicate events across process restarts. This may occur if the provider was recovered from
+	/// an old state (i.e., it hadn't been successfully persisted after processing pending events).
+	fn process_pending_events<H: Deref>(&self, handler: H) where H::Target: EventHandler;
+}
+
+/// A trait implemented for objects handling events from [`EventsProvider`].
+pub trait EventHandler {
+	/// Handles the given [`Event`].
+	///
+	/// See [`EventsProvider::process_pending_events`] for details that must be considered when
```

Comment:
Could point to the overall trait docs since those seem to have considerations (like the reentrancy) that could be pertinent as well